group/assignment5/Graph.java

```diff
@@ -0,0 +1,218 @@
+            } else if (childVertex.getState() == VertexState.UNVISITED) {
+                childVertex.setState(VertexState.VISITING);
+                dfsRecursive(childVertex, alphabet);
+                alphabet.add(childVertex.getValue());
+                childVertex.setState(VertexState.VISITED);
+            }
+        }
+    }
+
+    private void dfsForAllTopologicalOrders(List<T> alphabet, List<List<T>> alphabets) {
+        if (alphabet.size() == vertecies.size()) {
+            alphabets.add(new ArrayList<>(alphabet));
+        }
+        for (Vertex vertex : vertecies.values()) {
+            if (vertex.getState() == VertexState.UNVISITED && vertex.getIndegrees().size() == 0) {
+                vertex.getOutdegrees().forEach(child -> child.getIndegrees().remove(vertex));
+                alphabet.add(vertex.getValue());
+                vertex.setState(VertexState.VISITED);
+                dfsForAllTopologicalOrders(alphabet, alphabets);
+                vertex.setState(VertexState.UNVISITED);
+                alphabet.remove(alphabet.size() - 1);
+                vertex.getOutdegrees().forEach(child -> child.getIndegrees().add(vertex));
+            }
+        }
+    }
+
+    public Set<Constraint<T>> findCircuits() {
+        Set<Constraint<T>> constraints = new HashSet<>();
+        for (Vertex vertex : vertecies.values()) {
+            if (vertex.state == VertexState.UNVISITED) {
+                vertex.state = VertexState.VISITING;
+                dfsWithCircuitCheckRecursive(vertex, null, constraints);
+                vertex.state = VertexState.VISITED;
+            }
+        }
+        vertecies.values().forEach(e -> e.setState(VertexState.UNVISITED));
+        return constraints;
+    }
+
+    private void dfsWithCircuitCheckRecursive(Vertex current, Vertex parent, Set<Constraint<T>> constraints) {
+        for (Vertex childVertex : vertecies.get(current.getValue()).getOutdegrees()) {
+            if (childVertex.state == VertexState.VISITING) {
+                 constraints.add(new Constraint<>(parent.value, current.value));
+                 childVertex.state = VertexState.VISITED;
+            } else if (childVertex.state == VertexState.UNVISITED) {
+                childVertex.state = VertexState.VISITING;
+                dfsWithCircuitCheckRecursive(childVertex, current, constraints);
+                childVertex.state = VertexState.VISITED;
+            }
+        }
+    }
+
+    /**
+     * This class represents a vertex of a graph.
+     * Each vertex can be described by its character-value,
+     * indegree vertecies (list of vertecies that can direct you to the current vertex)
+     * and outdegree vertecies (list of vertecies that can be achieved from the current one)
+     * Vertex state is a special state to track the state of vertex during depth first search
+     * in a graph.
+     */
+    public class Vertex {
+        private T value;
+        private List<Vertex> indegrees;

```

Comment:
Maybe it would be better to use a TreeSet to enforce a sorted collection as well as no duplicate values?

Reply:
Right!

Reply:
Done using Hashset - the collection in and out degrees don't need to be sorted